### ISSUE TITLE
Error inlining limitGradients

### ISSUE THREAD

**Issue by SergioMartin86** [SergioMartin86/bamboo6#44]
This bug causes ROSE to avoid inlining limitGradients, which has MPI calls in its call tree. Although this doesn't prevent a successful translation, it affects compilation. If one solved the missing declaration and allows compilation to complete, the problem will still be far from solved. During runtime, actual MPI calls will be performed (not handled by tarragon) and probably a plethora of errors will occur.

I found the problem, though. This happens because of the enum argument type "limtype". Since the type is not explicitly specified (although we know that gcc assumes unsigned int), there's no way for ROSE to match it with its function definition in limiters.c (even though ROSE knows that the type name coincides... which is strange).

I want to try to change the definition of the typedef enum in mylimiters.h to see if I can explicitly define a type for it. If not, we will have to force a int type in the code (defintion and all calls to limitGradients).

**Comment by scottb5612** [comment]
Ugh.
There is an easy way around this one, since limtype (the enum type)
is used in only a few places.
But, are there any other places where enums will bite us?
They are used in many places in the cart3d libraries, too (limtype is used there, e.g. see SolverInfo.h)


" found the problem, though. This happens because of the enum argument type "limtype". "


So what you are saying is that ROSE expands enum into an unsigned int?
Scott

**Comment by SergioMartin86** [comment]
I have changed the enum into an int type and the problem persists, also tried removing the const modifiers. My official position now is that I have absolutely no clue what is causing the problem... but I'll figure it out.

**Comment by SergioMartin86** [comment]
Ok, Ive found the error and how to solve it:

Turns out there's more than one declaration for limitGradients. In particular, the problem is being caused because the function call expression:
> limitGradients(p_g, whichLimiter,level, glbls);

Has several different declarations for limitGradients all over the place, at least two of them being valid:
> void limitGradients(const p_tsGrid p_g, const limittype whichLimiter, const int level, const gstore *glbls); 
> void limitGradients(p_tsGrid p_g, limittype whichLimiter, int level,gstore *glbls); 

While the definition of the function validates only one of the declarations:
> void limitGradients(p_tsGrid p_g, int whichLimiter, int level,gstore *glbls) {blah blah blah...}

Turns out that ROSE matches the call expression with the first valid function declaration it finds (both are valid!). In our case, ROSE matches the call expression to the declaration with all the const modifiers. Although this is valid, since the Bamboo inliner is unable to find a definition for that declaration, it doesn't even evaluate it for inlining.

Solution is to either:
- Remove the inconsistent declaration (this is the correct solution)
- Put the correct declaration first, so ROSE will use it instead of the consts one (this is the ugliest solution ever).

But first, and more important, we need to figure out some way to collaborate over the cartd3d code without making it public (one alternative is to pay for the github service and making bamboo private). This problem reveals that there are problems with cart3d that are not fault of ROSE/Bamboo and we need to polish it before a successful translation can be done.

I guess we can define this over the phone, but if you read this msg before, you'll be better informed about the situation.

**Comment by scottb5612** [comment]
Thanks for digging into this.  I think that ROSE should really be reporting this sort of inconsistency,
but since it is not, we are going to need a tool to check all of cart3d to find any other places where this sort of thing is happening. I know at some point that Tan removed const qualifiers form the code, but in looking at the source that was provided to me originally, I can see that these inconsistencies were present already. I think it would be naive to think that we won't find any others.
Can you write a tool to do this? It will save us a lot of bellyaches.

**Comment by scottb5612** [comment]
I've thought about this a lot and  I think you've gotten to the bottom of something that I could  notfigure out without some detailed knowledge of ROSE. This business of types is messy and we need to think hard about it.  Is this related to the **typedef struct** problem is described earlier?

* This bug causes ROSE to avoid inlining limitGradients, which has MPI calls in its call tree. Although this doesn't prevent a successful translation, it affects compilation. If one solved the missing declaration and allows compilation to complete, the problem will still be far from solved. During runtime, actual MPI calls will be performed (not handled by tarragon) and probably a plethora of errors will occur. *

* I found the problem, though. This happens because of the enum argument type "limtype". Since the type is not explicitly specified (although we know that gcc assumes unsigned int), there's no way for ROSE to match it with its function definition in limiters.c (even though ROSE knows that the type name coincides... which is strange). *

**Comment by SergioMartin86** [comment]
No, that was my first hypothesis, but turned out to be incorrect. ROSE/Bamboo are fine with typedefs/enums as long as you don't use const modifiers. 

The big culprit for most of the annoying bugs we've been having is the use of const and the inconsistent headers. Like I've said to you once, there is absolutely no functional need to use const modifiers. Const is a nice feature of the language if used correctly, but that is not usually the case.

I really think the code in Cart3D needs to be normalized before we even think of debugging it. This involves:
- Reducing Cart3D to its bare bones. That is: removing all files that are not strictly needed for compilation/link/run.
- Have all code files (and nothing but code files) in a /src folder. All const modifiers need to be removed. Also a makefile, obviously.
- Have all header files in a /include folder. With this condition: there has to be one header file per source file, containing the function declarations (copy and pasted from their definitions) and relevant defines.
- Have a root / folder containing a Makefile automatizing translation/compilation/linking of both the normal variant of cart3D and the bamboo variant. Also, a ./run_test.sh file to run both versions for comparison.
